clang/lib/Sema/Sema.cpp

```diff
@@ -1761,8 +1761,15 @@ class DeferredDiagnosticsEmitter
   // Emit any deferred diagnostics for FD
   void emitDeferredDiags(FunctionDecl *FD, bool ShowCallStack) {
     auto It = S.DeviceDeferredDiags.find(FD);
-    if (It == S.DeviceDeferredDiags.end())
+    if (It == S.DeviceDeferredDiags.end()) {
```

Comment:
It is strange that we are having such troubles with template functions.
That seems to be a common problem among everyone who uses deferred diagnostics, see simple OpenMP example - https://godbolt.org/z/rKT4c8hbM .
The fix seems logical to me, but I'm not sure we are meant to this thing here. I'd like to double check with my own eyes what is happening here before suggesting any other solutions. I'll get back to this tomorrow.

BTW, no matter which solution we will choose, this appears to be a common non-SYCL-only problem, so it is probably better to be upstreamed.

Reply:
So, I've spent some time under debugger and it right now nothing that looks better than your patch comes to my mind. The diagnostic is issued quite early, the global variable doesn't depend on template parameters, so `DiagnoseUseOfDecl` (the function where the diagnostic is issued) is not called for the second time once the template is instantiated.
I suggest to try a couple of more complicated tests with bigger level of nested templates called from device context and then try to bring this patch to community. As I said, this problem likely affects everyone who uses deferred diagnostics (see the godbolt link above), so people might be interested in it. Also the change affects non-sycl code.

Reply:
Thanks, will take that approach.   Appreciate your looking into this.

Reply:
Sorry to revive this thread this late.   I looked into adding this patch to the community and there are a couple of issues with directly upstreaming this:
- the first is that we do a lot of diagnosing in ``DiagnoseUseOfDecl`` for SYCL, and there aren't many equivalents for say, OpenMP in that routine.   And it is the initial diagnosis here that gets the template declaration in the deferred diagnostics stack.  So the absence of this leads to this fix not being very useful for the general community.
- the second is that the getEmissionStatus routine for non-SYCL compilations have a check for template declaration and simply discard the diagnostic.  We use the same routine, but SYCL chooses to Defer even if it is in a template.

It is the combination of these two: i.e. opportunity to diagnose SYCL violations in ``DiagnoseUseOfDecl`` followed by the ability to defer the diagnostic when emitting (instead of discarding it) in a template context that makes my fix of checking for the template declaration even viable.

@Fznamznon, what are your thoughts given this?   May I check my changes into sycl here?

Reply:
> the first is that we do a lot of diagnosing in DiagnoseUseOfDecl for SYCL, and there aren't many equivalents for say, OpenMP in that routine. And it is the initial diagnosis here that gets the template declaration in the deferred diagnostics stack. So the absence of this leads to this fix not being very useful for the general community.

I understand your concern, If I remember correct we didn't upstream much SYCL diagnostics in `DiagnoseUseOfDecl`. I don't think the problem is only with `DiagnoseUseOfDecl` though, see OpenMP example https://godbolt.org/z/rKT4c8hbM that I proposed earlier. I guess it is just depends on the fact if the restricted code depends on the template parameters. If not, then only non-instantiated template function will get into deferred diagnostics callgraph, so we miss a diagnostic.

> the second is that the getEmissionStatus routine for non-SYCL compilations have a check for template declaration and simply discard the diagnostic. We use the same routine, but SYCL chooses to Defer even if it is in a template.

Oops, It seems I've removed this check in https://github.com/intel/llvm/pull/5195/files , and it didn't really affect anything. But It may be worth to double check. Also there was a check on sycl_device and sycl_kernel attributes, so when we have a template context not marked with there attributes - the problem with missing diagnostic is still here.

> what are your thoughts given this? May I check my changes into sycl here?

I understand that we'd like to fix a diagnostic that comes from `DiagnoseUseOfDecl` and it is not yet upstreamed, so It might make sense to check it here, however it seems we are not sure whether it is a good/correct fix and the problem is not SYCL-specific, so we could try and upstream it using simple OpenMP example I provided above and some simple SYCL diagnostic. Anyway in no case I didn't want to block this patch, I just saw an opportunity to get help from community and highlighted it.

Reply:
The code-flow inside ``getEmissionStatus`` is:

  ```
  if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLKernelAttr>())
    return FunctionEmissionStatus::Emitted;
 
  if (FD->isDependentContext())
    return FunctionEmissionStatus::TemplateDiscarded;

  // OpenMP logic follows
 
  ...
```
SYCL decides to emit if it is a kernel function before it checks if it is a temple; if I remove this check or move it after the OpenMP logic, we generate a lot more false diagnostics for OpenMP.  i.e. we issue diagnostics for function templates inside OpenMP regions, even if those functions are not instantiated.   This is even without the changes in Sema.cpp that is under review here.

I think the issue here is that the diagnostics mechanism for SYCL and OpenMP (I have not looked at CUDA) are different enough that, even though my change for SYCL appears generic, it is not helpful to other components.  On the other hand, some trivial attempts to make it work for OpenMP by changing other logic, seems to be making it worse for those components.

Reply:
Ok, I have checked my OpenMP example with your change and it still doesn't work. Then I agree that this particular change doesn't help to other components and we can commit it here since it fixes diagnostic that is not yet upstreamed. Sorry for the confusion and thank you for your patience.
Please note that https://github.com/intel/llvm/pull/5195 have changed code-flow inside `getEmissionStatus` to:
```
  // Templates are emitted when they're instantiated.
  if (FD->isDependentContext())
    return FunctionEmissionStatus::TemplateDiscarded;

  if (LangOpts.SYCLIsDevice &&
      (FD->hasAttr<SYCLDeviceAttr>() || FD->hasAttr<SYCLKernelAttr>()))
    return FunctionEmissionStatus::Emitted;

// OpenMP logic follows
```

Reply:
@Fznamznon, no worries about the confusion, and thank YOU for the persistence.   It would have been nice to have some of our changes help the community, and perhaps we will be able to do that in this area when we converge in the future.

Thanks again for pointing out your change in #5195 as well as your explanation here: https://github.com/intel/llvm/pull/5195/files#r773379931

The ordering does not affect SYCL today because ``SYCLDiagIfDeviceCode`` does not check ``getEmissionStatus``, similar to how ``CUDADiagIfDeviceCode`` or ``OpenMPDiagIfDeviceCode`` does.  I think this difference too should be somehow be resolved and perhaps this check logic moved to targetDiag instead of to the underlying ``...DiagIf...`` routines.

I will upload a new patch with your previous minor code-review comment addressed.